File: app/src/main/java/com/leeeyou/zhihudaily/view/ZhiHuDailyFragment.java
package com.leeeyou.zhihudaily.view;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.leeeyou.BaseFragment;
import com.leeeyou.R;
import com.leeeyou.zhihudaily.model.ZhiHuDailyRepositoryKt;
import com.leeeyou.zhihudaily.model.bean.ZhiHuDaily;
import com.leeeyou.zhihudaily.model.bean.ZhiHuDailyItem;

import org.joda.time.DateTime;

import java.util.List;

import in.srain.cube.views.ptr.PtrClassicFrameLayout;
import in.srain.cube.views.ptr.PtrDefaultHandler;
import in.srain.cube.views.ptr.PtrFrameLayout;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

/**
 * ClassName:   ZhiHuDailyFragment
 * Description: ZhiHuDailyFragment
 * <p>
 * Author:      leeeyou
 * Date:        2017/4/24 13:46
 */
public class ZhiHuDailyFragment extends BaseFragment {

    private RecyclerView mRecyclerView;
    private PtrClassicFrameLayout mPtrFrame;
    private ZhiHuDailyAdapter mAdapter;

    //when loading more, up to the date of the data can be loaded
    private String[] dateList = new String[7];
    private int mDatePosition = 0;
    private int mMostDate = 7;

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View inflate = inflater.inflate(R.layout.activity_zhihu_daily, container, false);
        mRecyclerView = inflate.findViewById(R.id.recyclerViewRiBao);
        mPtrFrame = inflate.findViewById(R.id.ptrFrame);
        return inflate;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initDateList();
        initView();
        fetchZhiHuDailyList();
    }

    private void initView() {
        mPtrFrame.disableWhenHorizontalMove(true);
        mPtrFrame.setPtrHandler(new PtrDefaultHandler() {
            @Override
            public void onRefreshBegin(PtrFrameLayout frame) {
                mDatePosition = 0;
                fetchZhiHuDailyList();
            }

            @Override
            public boolean checkCanDoRefresh(PtrFrameLayout frame, View content, View header) {
                return !mRecyclerView.canScrollVertically(-1);
            }
        });

        initAdapter();
    }

    private void initAdapter() {
        mAdapter = new ZhiHuDailyAdapter(R.layout.item_zhihu_daily, null);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
        mRecyclerView.setAdapter(mAdapter);

        mAdapter.openLoadAnimation(BaseQuickAdapter.SCALEIN);
        mAdapter.setOnLoadMoreListener(new BaseQuickAdapter.RequestLoadMoreListener() {
            @Override
            public void onLoadMoreRequested() {
                if (mDatePosition < mMostDate - 1) {
                    fetchZhiHuDailyList();
                } else {
                    mAdapter.loadMoreEnd();
                }
            }
        }, mRecyclerView);
    }

    private void fetchZhiHuDailyList() {
        ZhiHuDailyRepositoryKt
                .fetchZhiHuDailyListByDate(dateList[++mDatePosition])
                .subscribeOn(Schedulers.newThread())
                .filter(new Func1<ZhiHuDaily, Boolean>() {
                    @Override
                    public Boolean call(ZhiHuDaily riBao) {
                        StringBuilder sb = new StringBuilder();
                        char[] chars = riBao.getDate().toCharArray();
                        for (int i = 0; i < chars.length; i++) {
                            if (i == 4 || i == 6) {
                                sb.append("-");
                            }
                            sb.append(chars[i]);
                        }

                        List<ZhiHuDailyItem> stories = riBao.getStories();
                        for (ZhiHuDailyItem item : stories) {
                            item.setDate(sb.toString());
                        }

                        return true;
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<ZhiHuDaily>() {
                    @Override
                    public void onCompleted() {
                        mAdapter.loadMoreComplete();
                    }

                    @Override
                    public void onError(Throwable e) {
                        mPtrFrame.refreshComplete();
                        e.printStackTrace();
                        com.leeeyou.util.T.showShort(getActivity(), "出错了:" + e.getMessage());
                    }

                    @Override
                    public void onNext(ZhiHuDaily zhiHuDaily) {
                        mPtrFrame.refreshComplete();
                        setDataToAdapter(zhiHuDaily.getStories());
                    }
                });
    }

    private void setDataToAdapter(@NonNull List<ZhiHuDailyItem> zhiHuDailyItemList) {
        if (mDatePosition == 0) {
            mAdapter.setNewData(zhiHuDailyItemList);
        } else {
            mAdapter.addData(zhiHuDailyItemList);
        }
    }

    private void initDateList() {
        DateTime mDateTime = DateTime.now();
        for (int i = 0; i < mMostDate; i++) {
            DateTime tempDateTime = mDateTime.minusDays(i);
            dateList[i] = String.valueOf(tempDateTime.getYear()) +
                    (tempDateTime.getMonthOfYear() < 10 ? "0" + tempDateTime.getMonthOfYear() : tempDateTime.getMonthOfYear()) +
                    (tempDateTime.getDayOfMonth() < 10 ? "0" + tempDateTime.getDayOfMonth() : tempDateTime.getDayOfMonth());
        }
    }
}

File: app/src/main/java/com/leeeyou/zhihudaily/view/ZhiHuDailyFragment.kt
package com.leeeyou.zhihudaily.view

import `in`.srain.cube.views.ptr.PtrDefaultHandler
import `in`.srain.cube.views.ptr.PtrFrameLayout
import `in`.srain.cube.views.ptr.header.StoreHouseHeader
import `in`.srain.cube.views.ptr.util.PtrLocalDisplay.dp2px
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.chad.library.adapter.base.BaseQuickAdapter
import com.leeeyou.R
import com.leeeyou.manager.BaseFragment
import com.leeeyou.manager.MyLoadMoreView
import com.leeeyou.util.startBrowserActivity
import com.leeeyou.widget.LoadingDialog
import com.leeeyou.zhihudaily.model.bean.ZhiHuDaily
import com.leeeyou.zhihudaily.model.bean.ZhiHuDailyDetail
import com.leeeyou.zhihudaily.model.bean.ZhiHuDailyItem
import com.leeeyou.zhihudaily.model.fetchZhiHuDailyDetailById
import com.leeeyou.zhihudaily.model.fetchZhiHuDailyListByDate
import kotlinx.android.synthetic.main.activity_zhihu_daily.*
import org.joda.time.DateTime
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import timber.log.Timber

class ZhiHuDailyFragment : BaseFragment() {
    private val mMostDate = 7
    private val mDateList = Array(7) { "" }
    private var mDatePosition = 0
    private lateinit var mAdapter: ZhiHuDailyAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.activity_zhihu_daily, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initDataList()
        initView()
        fetchZhiHuDailyList(mDateList[0])
    }

    private fun updateAdapter(zhiHuDailyItemList: List<ZhiHuDailyItem>) {
        if (mDatePosition == 0) {
            mAdapter.setNewData(zhiHuDailyItemList)
        } else {
            mAdapter.addData(zhiHuDailyItemList)
        }
    }

    private fun initView() {
        initPtr()
        initAdapter()
    }

    private fun initPtr() {
        initHeadView()
        ptrFrameOfZhiHuDaily.disableWhenHorizontalMove(true)
        ptrFrameOfZhiHuDaily.setPtrHandler(object : PtrDefaultHandler() {
            override fun onRefreshBegin(frame: PtrFrameLayout?) {
                mDatePosition = 0
                fetchZhiHuDailyList(mDateList[mDatePosition])
            }

            override fun checkCanDoRefresh(frame: PtrFrameLayout?, content: View?, header: View?): Boolean {
                return !recyclerViewRiBao.canScrollVertically(-1) && (activity as com.leeeyou.IndexActivity).appBarLayoutVerticalOffset >= 0
            }
        })
    }

    private fun initHeadView() {
        val header = StoreHouseHeader(context)
        header.setTextColor(resources.getColor(R.color.colorTxtSelected))
        header.setPadding(0, dp2px(15f), 0, 0)
        header.initWithString("ZhiHu Daily", 15)
        ptrFrameOfZhiHuDaily.headerView = header
        ptrFrameOfZhiHuDaily.addPtrUIHandler(header)
    }

    private fun initAdapter() {
        mAdapter = ZhiHuDailyAdapter(R.layout.item_zhihu_daily, null)
        mAdapter.openLoadAnimation(BaseQuickAdapter.SCALEIN)
        mAdapter.setLoadMoreView(MyLoadMoreView())
        mAdapter.setOnItemChildClickListener { adapter, _, position ->

            var loadingDialog: LoadingDialog? = null
            this@ZhiHuDailyFragment.context?.also {
                loadingDialog = LoadingDialog(it)
                loadingDialog?.show()
            }

            val item: ZhiHuDailyItem? = adapter.getItem(position) as ZhiHuDailyItem
            item?.also {
                fetchZhiHuDailyDetailById(it.id)
                        .subscribeOn(Schedulers.newThread())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(object : Subscriber<ZhiHuDailyDetail>() {
                            override fun onCompleted() {
                                Timber.i("fetchZhiHuDailyDetailById onCompleted")
                                loadingDialog?.dismiss()
                            }

                            override fun onError(e: Throwable) {
                                Timber.e(e)
                                loadingDialog?.dismiss()
                            }

                            override fun onNext(storyDetail: ZhiHuDailyDetail) {
                                Timber.i("fetchZhiHuDailyDetailById onNext")
                                loadingDialog?.dismiss()
                                startBrowserActivity(context!!, storyDetail.share_url, storyDetail.title)
                            }
                        })
            }
        }
        mAdapter.setOnLoadMoreListener({
            if (mDatePosition < mMostDate - 1) {
                fetchZhiHuDailyList(mDateList[++mDatePosition])
            } else {
                mAdapter.loadMoreEnd()
            }
        }, recyclerViewRiBao)

        recyclerViewRiBao.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        recyclerViewRiBao.adapter = mAdapter
    }

    private fun initDataList() {
        val now = DateTime.now().plusDays(1)
        for (i in 0 until mMostDate) {
            val tempDateTime = now.minusDays(i)
            val year = tempDateTime.year.toString()
            val month = if (tempDateTime.monthOfYear < 10) "0" + tempDateTime.monthOfYear else tempDateTime.monthOfYear.toString()
            val day = if (tempDateTime.dayOfMonth < 10) "0" + tempDateTime.dayOfMonth else tempDateTime.dayOfMonth.toString()
            mDateList[i] = year + month + day
        }
    }

    private fun fetchZhiHuDailyList(date: String) {
        fetchZhiHuDailyListByDate(date)
                .subscribeOn(Schedulers.newThread())
                .filter { zhiHuDaily ->
                    for (item in zhiHuDaily.stories) {
                        item.date = StringBuilder(zhiHuDaily.date)
                                .insert(4, "-")
                                .insert(7, "-")
                                .toString()
                    }
                    true
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Subscriber<ZhiHuDaily>() {
                    override fun onNext(item: ZhiHuDaily?) {
                        ptrFrameOfZhiHuDaily?.refreshComplete()
                        if (item != null) {
                            updateAdapter(item.stories)
                        }
                    }

                    override fun onCompleted() {
                        mAdapter.loadMoreComplete()
                    }

                    override fun onError(e: Throwable?) {
                        ptrFrameOfZhiHuDaily?.refreshComplete()
                        e?.printStackTrace()
                        Timber.e(e)
                    }
                })
    }
}